libs/filamat/src/MaterialBuilder.cpp

```diff
@@ -1006,6 +1027,69 @@ Package MaterialBuilder::build(JobSystem& jobSystem) noexcept {
     return package;
 }
 
+using namespace backend;
+static const char* to_string(ShaderStageFlags stageFlags) noexcept {
+    switch (stageFlags) {
+        case ShaderStageFlags::NONE:                    return "{ }";
+        case ShaderStageFlags::VERTEX:                  return "{ vertex }";
+        case ShaderStageFlags::FRAGMENT:                return "{ fragment }";
+        case ShaderStageFlags::ALL_SHADER_STAGE_FLAGS:  return "{ vertex | fragment }";
+    }
+    return nullptr;
+}
+
+bool MaterialBuilder::checkMaterialLevelFeatures(MaterialInfo const& info) const noexcept {
+
+    auto logSamplerOverflow = [](SamplerInterfaceBlock const& sib) {
+        auto const& samplers = sib.getSamplerInfoList();
+        auto const* stage = to_string(sib.getStageFlags());
+        for (auto const& sampler: samplers) {
+            slog.e << "\"" << sampler.name.c_str() << "\" "
+                    << Enums::toString(sampler.type).c_str() << " " << stage << '\n';
+        }
+        flush(slog.e);
+    };
+
+    const auto userSamplerCount = info.sib.getSize();
+    switch (info.featureLevel) {
+        case FeatureLevel::FEATURE_LEVEL_1: {
+            // TODO: we need constants somewhere for these values
```

Comment:
Could we use `PerViewSib::SAMPLER_COUNT` to derive this? Is this number still accurate if morphing is used?

Reply:
We could say that it's e.g.: `16 - "what we need in filament"` but I think it would be wrong to do it this way. I think it should be the hardcoded number that is in our (non existant) docs. Because we could make mistakes in `PerViewSib` for instance.

This number should be accurate "by definition". In practice it is because morphing doesn't count against the "fragment" sampler count.